### PR TITLE
Relax RV requirements

### PR DESCRIPTION
**Pull request by lavalamp**
#### What type of PR is this?

/kind bug
/kind documentation

/kind api-change

#### What this PR does / why we need it:

Relaxes rules around what clients can do with resource version.

#### Which issue(s) this PR fixes:

Partly addresses #112684 

#### Does this PR introduce a user-facing change?
```release-note
NONE
```

### CODE REVIEW

**Comment by lavalamp** [comment]
(I didn't bother to update generated files because I expect a bunch of comments)

File: staging/src/k8s.io/apimachinery/pkg/apis/meta/v1/types.go [hunk 1]
```diff
@@ -158,15 +158,24 @@ type ObjectMeta struct {
...
-	// An opaque value that represents the internal version of this object that can
+	// A value that represents the internal version of this object that can
 	// be used by clients to determine when objects have changed. May be used for optimistic
 	// concurrency, change detection, and the watch operation on a resource or set of resources.
-	// Clients must treat these values as opaque and passed unmodified back to the server.
-	// They may only be valid for a particular resource or set of resources.
+	//
+	// This value is a string containing a base-10 encoding of a 63 bit positive integer.
+	// Clients may compare this value for equality between objects. If parsed as above, clients
+	// may compare for recency between two objects of the same API Group and Kind.
```
Comment (aojea) [inline staging/src/k8s.io/apimachinery/pkg/apis/meta/v1/types.go:167]: do we have a test that assert we keep this promise 😄 

I think @stevekuznetsov was doing some tests related to this https://github.com/kubernetes/kubernetes/pull/109730, don't know if those can be adapted/reused

**Comment by deads2k** [comment]
the doc lgtm

File: staging/src/k8s.io/apimachinery/pkg/apis/meta/v1/types.go (hunk 1 above)
Comment (stevekuznetsov) [reply staging/src/k8s.io/apimachinery/pkg/apis/meta/v1/types.go:167]: I think the only current break we know of is events with expiring TTL.

File: staging/src/k8s.io/apimachinery/pkg/apis/meta/v1/types.go [hunk 2]
```diff
@@ -158,15 +158,24 @@ type ObjectMeta struct {
...
 	UID types.UID `json:"uid,omitempty" protobuf:"bytes,5,opt,name=uid,casttype=k8s.io/kubernetes/pkg/types.UID"`
 
-	// An opaque value that represents the internal version of this object that can
+	// A value that represents the internal version of this object that can
 	// be used by clients to determine when objects have changed. May be used for optimistic
 	// concurrency, change detection, and the watch operation on a resource or set of resources.
-	// Clients must treat these values as opaque and passed unmodified back to the server.
-	// They may only be valid for a particular resource or set of resources.
+	//
+	// This value is a string containing a base-10 encoding of a 63 bit positive integer.
```
Comment (lavalamp) [inline staging/src/k8s.io/apimachinery/pkg/apis/meta/v1/types.go:165]: The status of this PR is that we're waiting for the SIG to hear the case for making this arbitrary-length.

File: staging/src/k8s.io/apimachinery/pkg/apis/meta/v1/types.go (hunk 1 above)
Comment (wojtek-t) [reply staging/src/k8s.io/apimachinery/pkg/apis/meta/v1/types.go:167]: In particular, we're (unfortunately) relying on that in watchcache, so it's heavily battle-tested.

Yes - the only case I'm aware when this breaks is https://github.com/kubernetes/kubernetes/issues/110210

File: staging/src/k8s.io/apimachinery/pkg/apis/meta/v1/types.go [hunk 3]
```diff
@@ -158,15 +158,24 @@ type ObjectMeta struct {
...
+	// Clients may compare this value for equality between objects. If parsed as above, clients
+	// may compare for recency between two objects of the same API Group and Kind.
+	//
+	// It is undefined to compare this value across objects of differing API Group or Kind.
+	// Clients may not do that.
```
Comment (wojtek-t) [inline staging/src/k8s.io/apimachinery/pkg/apis/meta/v1/types.go:170]: or "must not do that"?

**Comment by smarterclayton** [comment]
One concern I have here is that this changes a field that is under conformance, and broadening it’s definition makes currently conformant distributions (they are passing).  Wearing my conformance hat, i would probably block this from being part of conformance because of that impact (changing what is allowed).  I’d like to get some extra opinions there before this merges for sure, and honestly this is a KEP, not a PR.

/hold

**Comment by lavalamp** [comment]
@smarterclayton 

This was discussed during the API machinery meeting today, we haven't been doing this in secret, and it's been deliberately very slow to give plenty of time for feedback. I do not think we need KEPs for documentation changes that merely admit the way things really are (this change).

I don't agree at all that this imposes some new conformance condition on users? This represents the least obligation possible to place on users (none)? The other option (per-object clock) would have required a KEP, but this option is literally changing nothing in practice, it just stops lying to people in the documentation. 

At any rate, if you don't like this, you will probably like even less that the SIG decided earlier today to let @stevekuznetsov make a version of this change that moves to arbitrary length numbers, which does have an impact against clients (I argued against but without success).

**Comment by smarterclayton** [comment]
> I do not think we need KEPs for documentation changes that merely admit the way things really are (this change).

We are explicitly endorsing a tightening of requirements for the underlying storage and changing our documented api guarantees.  We’ve certainly spilled a KEPs worth of words describing the impact and considerations, and it certainly isn’t an easy discussion.  I don’t object to the intent specifically, I’m concerned about the process and impact.  Documenting via a KEP the precise intent and range of behavior allowed and our reasons for changing (and the implications) is what we reasonably ask for when we change the behavior of the system.  

If the position of api machinery is that the storage must follow a set of rules, it’s pretty obvious we should document and test those rules in a way that others can follow.

**Comment by smarterclayton** [comment]
> I don't agree at all that this imposes some new conformance condition on users? 

The issue that I was working through the implications of was:

1. decision to change our formal position (not enforced by code today) in a way that potentially breaks or at least requires changes to alternate backends (at somewhere between a low % and 0%)
2. if we change our design intent, we should be adding a test that verifies it
3. if we add such a test, would we allow it in conformance?  probably, since we're now saying the conformal definition of kube is that you emulate this set of etcd derived behaviors (we are formalizing our storage requirements, which should be a KEP, hence my other comment) that are client observable
4. However, conformance is focused very much on application portability - can a workload that works on one conformant distribution work on others.  In the spirit of conformance, the behavior we have today of people who parse the RV is "not part of conformance, and so you can be broken if you parse it" (rather than the position the sig is endorsing, which is "we only support etcd") 
5. if we added it in conformance, and it broke a distribution that was previously conformant, would we revert this change?

At that point we would have evidence of use and evidence of regression so probably yes - conformance defines the behavior clients accept, and if an implementation was currently conformant and was no longer due to this change, it's an api change and should follow our API rules (both on the process and outcome side). 

Most of this discussion hinges on "is anyone practically broken if we change the definition" - we can imagine someone who might be broken, but we don't have vocal participation in the sig yet of anyone who is broken.  I'm concerned that is a very low bar ("if you don't show up, we can break you"), so the mentioned conformance thread is an attempt to gather data from conformant distributions (people who have an incentive to describe themselves as conformant if they do change out their data layer) in an attempt to bound the uncertainty a bit prior to such a change.

My general position is that tightening what the definition of Kube is is KEP-worthy in general and demands some level of dilligence in driving communication about the potential change in the ecosystem.  We've done about 50% of that, but I'd prefer all the ink that has been spilled arguing about this not be in vain and serve as a normative reference, and this doc change doesn't seem to accomplish that.

**Pull request closed without merging**